### DISCUSSION TITLE
TriggerDagRunOperator and ExternalTaskSensor with dynamic execution_date

### DISCUSSION THREAD

**Discussion by prithvi-git** [apache/airflow#36252] [Q&A]
Trying to trigger one dag multiple times with different configs using TriggerDagRunOperator and ExternalTaskSensor. However, TriggerDagRunOperator takes parent DAGs execution_date (logical_date) for execution and that just reruns same instance of triggered DAG instead of running new instance with new config. Sample code below.
Tried passing execution_date if different ways to TriggerDagRunOperator and ExternalTaskSensor but no luck.
I'm expecting a solution so I can trigger same DAG with different config.

# DAG - exec_master     
@task_group(group_id='refresh_pre-prod')
    def refresh_pre_prod():
        prod_to_pre_prod = TriggerDagRunOperator (
            task_id='prod_to_pre_prod',
            trigger_dag_id="util_clone_bq_env",
            execution_date='{{ dag_run.logical_date }}',
            conf={"src_project_id":"production",
                    "trg_project_id":"pre-production"}, 
            reset_dag_run = True)

        prod_to_pre_prod_sensor = ExternalTaskSensor(
            task_id='prod_to_pre_prod_sensor',
            external_dag_id='util_clone_bq_env',
            external_task_id='notify_completion',
            allowed_states=["success"],
            failed_states=["failed", "skipped", "upstream_failed"])

        prod_to_pre_prod >> prod_to_pre_prod_sensor

    @task_group(group_id='refresh_demo')
    def refresh_demo():
        prod_to_demo = TriggerDagRunOperator(
            task_id='prod_to_demo',
            trigger_dag_id="util_clone_bq_env",
            execution_date='{{ dag_run.logical_date }}',
            conf={"src_project_id":"production",
                    "trg_project_id":"demo1"},
            reset_dag_run = True)

        prod_to_demo_sensor = ExternalTaskSensor(
            task_id='prod_to_demo_sensor',
            external_dag_id='util_clone_bq_env',
            external_task_id='notify_completion',
            allowed_states=["success"],
            failed_states=["failed", "skipped", "upstream_failed"])

        prod_to_demo >> prod_to_demo_sensor
        
refresh_pre_prod() >> bq_refresh_demo()

Thanks,
Prithvi.

**Comment by josh-fell** [comment]
Hi @prithvi-git,

IIUC your main problem that the "prod_to_demo_sensor" task doesn't properly find the DAG run that's triggered by the "prod_to_demo"? If so, you could always have the "prod_to_demo" task continue to run until the triggered DAG is completed by setting [`wait_for_completion=True` on the TriggerDagRunOperator](https://registry.astronomer.io/providers/apache-airflow/versions/2.7.3/modules/TriggerDagRunOperator) task. That operator also has both `allowed_states` and `failed_states` too.

**Reply by prithvi-git** [reply]
Hi @josh-fell 

Thanks for your response. 

Above tasks are part of one master DAG (exec_master), from where I'm trying to trigger another DAG (util_clone_bq_env) with different conf values for pre-prod and demo projects and run them sequentially as refresh_pre_prod() >> bq_refresh_demo(). 
Expectation here is, it should run util_clone_bq_env DAG first with conf for pre_prod and then with conf from demo. However, since the execution_date is set to '{{ dag_run.logical_date }}' which is execution date of exec_master DAG, which remains same in both the triggers, it runs the util_clone_bq_env DAG for pre-prod, but for demo trigger it just reruns the execution instance of pre-prod.
Can you help me how to achieve this?

Also thanks for highlighting states parameters.

Best Regards,
Prithvi.

**Reply by josh-fell** [reply]
If the `execution_date` is the same, the [operator will reset and rerun the same DAG run because you have `reset_dag_run=True`](https://github.com/apache/airflow/blob/58886ca56201690737d1e0d9d720665e2192585f/airflow/operators/trigger_dagrun.py#L164C1-L187r). Otherwise, if you don't set the `execution_date` in the TriggerDagRunOperator, the [`execution_date` of the triggered DAG would be the current UTC time.](https://github.com/apache/airflow/blob/58886ca56201690737d1e0d9d720665e2192585f/airflow/operators/trigger_dagrun.py#L147C1-L152).

Additionally, from the `reset_dag_run` parameter description:
> Dag run conf is immutable and will not be reset on rerun of an existing dag run.

So it seems as though you might be better off _not_ setting `execution_date` on the TriggerDagRunOperator task and having the task wait for the completion of the triggered DAG run.

**Reply by prithvi-git** [reply]
Thanks @josh-fell 
I tried this. In this case, ExternalTaskSensor keeps running forever since it is poking to instance with execution_date as master DAGs execution_date (i.e. logical_date)
I tried execution_date_fn to pass current UTC time, but there is always a slight difference in time between TriggerDagRunOperator and ExternalTaskSensor. So ExternalTaskSensor keeps running forever.

**Reply by josh-fell** [reply]
Have you tried just having the `TriggerDagRunOperator` wait for completion?

> ... you could always have the "prod_to_demo" task continue to run until the triggered DAG is completed by setting [wait_for_completion=True on the TriggerDagRunOperator](https://registry.astronomer.io/providers/apache-airflow/versions/2.7.3/modules/TriggerDagRunOperator) task.

**Reply by prithvi-git** [reply]
@josh-fell 
"wait_for_completion=True" didn't help. prod_to_demo still triggers the same instance triggered for prod_to_pre_prod :(
Unless we are able to pass distinct logical_date in TriggerDagRunOperator and able to use them in respective ExternalTaskSensor tasks, it will not trigger other instance of triggered DAG (util_clone_bq_env)

**Reply by prithvi-git** [reply]
AIRFLOW_CTX_EXECUTION_DATE=2023-12-29T07:39:42.683753+00:00
It's same for both triggers, hence running same instance twice.

**Reply by raunakchopra** [reply]
Were you able to get around this issue by any chance? I am also here, struggling with the Poke forever issue..

**Reply by prithvi-git** [reply]
No.. This issue is not resolved yet. I had to create multiple DAGs with different names but exact same code and call them.
